File: ch1/pal_perm.cc
// Given a string, pal_perm checks to see if it is a permutation of a
//     palindrome.

#include <iostream>
#include <string>
#include <map>
#include <algorithm>

bool perm_pal(std::string &str) {
  // If the string is a palindrome, then all characters will be even, with
  //     possibly one odd
  int len = str.length();
  bool odd_char = false;

  std::map<char, int> charmap;

  // adding all characters to a map (also, make lowercase)
  for (int i = 0; i < len; i++) {
    if (str[i] == ' ') continue;

    if (str[i] >= 'A' && str[i] <= 'Z') {
      str[i] = str[i] - 'A' + 'a';
    }
    charmap[str[i]]++;
  }

  // checking for palindrome property
  for (auto &x: charmap) {
    if (x.second % 2 == 1 && odd_char == false) {
      odd_char = true;
    }

    else if (x.second % 2 == 1 && odd_char == true) {
      return false;
    }
  }

  return true;
}

int main() {
  std::string str1 = "Tact Coa";
  std::string str2 = "ABCCBA";
  std::string str3 = "ABC";
  std::cout << std::boolalpha << perm_pal(str1) << std::endl;
  std::cout << std::boolalpha << perm_pal(str2) << std::endl;
  std::cout << std::boolalpha << perm_pal(str3) << std::endl;
}

File: ch1/pal_perm.py
def pal_perm(str1):
    # palindrome must have 2 of each char plus potentially 1 of another
    charmap = {}
    fstr = str1.upper().replace(' ','')

    for char in fstr:
        val = charmap.get(char, 0)

        if val == 0:
            charmap.update({char: 1})

        elif val == 2:
            return False

        else: 
            charmap.update({char: val+1})

    # iterate through charmap
    single = False
    for k, v in charmap.items():
        if v == 1 and not single:
            single = True
        elif v == 1 and single:
            return False
    
    return True

def main():
    str1 = "Tact Coa"
    str2 = "ABCCBA"
    str3 = "ABCDEF"

    print(pal_perm(str1))
    print(pal_perm(str2))
    print(pal_perm(str3))

if __name__ == "__main__":
    main()